### DISCUSSION TITLE
Enable pop-up Notification for app running in foreground

### DISCUSSION THREAD

**Discussion by rawatnaresh** [invertase/react-native-firebase#4187] [Q&A] [answered]
I could see a popup notification when my app is inactive but I'm trying to achieve the same behavior when my app is running in foreground mode.

Handling background state messages
```
// This will only show a popup when my app is running in background
messaging().setBackgroundMessageHandler(async remoteMessage => {});
```
Handling foreground state messages
```
const unsubscribe = messaging().onMessage(async remoteMessage => {
  Alert.alert('A new FCM message arrived!', JSON.stringify(remoteMessage));
  // instead of alert I want to show a pop up notification
});
return unsubscribe;
```

I could achieve this by using [notifee](https://notifee.app/react-native/docs/displaying-a-notification)

```
const unsubscribe = messaging().onMessage(async remoteMessage => {
  // Display a notification
  await notifee.displayNotification({
    title: 'Notification Title',
    body: 'Main body content of the notification',
    android: {
      channelId,
    },
   });
});
return unsubscribe;
```

But is there a way I can do this without using any other packages?

**Comment by mikehardy** [comment, marked as answer]
No, there is no way to do this without a full featured 3rd party notification package.

There are 3 packages we are aware of that may serve as full-featured notification packages:

- Notifee: https://notifee.app
- React Native Push Notification: https://github.com/zo0r/react-native-push-notification
- React Native Notifications: https://github.com/wix/react-native-notifications

All have reported success.

**Reply by wesvh** [reply]
On expo, have alternative?

**Reply by Muhammed-Walied** [reply]
what about expo ?

**Reply by mimo-10** [reply]
an expo there is alternative, you could trigger a local notification using `scheduleNotificationAsync`

**Comment by simonbengtsson** [comment]
I assume the above answer is for Android or outdated. For iOS you can control the foreground behavior with the `messaging_ios_foreground_presentation_options` in firebase.json https://rnfirebase.io/messaging/usage#foreground-presentation-options-ios.

**Comment by BlagojeV93** [comment]
Feels unfortunate that only in case of the Android foreground notifications, there's a need for external notification library

**Reply by mikehardy** [reply]
But what about the case of all those fancy iOS permissions which change every version, and knowing their status?
The API surface area surrounding device-local permissions is simply too large to host adapters for in this repo
